FreeRTOS/Demo/CORTEX_MPU_M4_MPS2_QEMU_GCC/startup.c

```diff
@@ -0,0 +1,253 @@
+        "Infinite_Loop:\n"
+        "    b  Infinite_Loop\n"
+        ".size  Default_Handler, .-Default_Handler\n"
+        ".align 4\n"
+        "NVIC_INT_CTRL_CONST: .word 0xe000ed04\n"
+    );
+}
+static void HardFault_Handler( void ) __attribute__( ( naked ) );
+void HardFault_Handler( void )
+{
+    __asm volatile
+    (
+        " tst lr, #4                                                \n"
+        " ite eq                                                    \n"
+        " mrseq r0, msp                                             \n"
+        " mrsne r0, psp                                             \n"
+        " ldr r1, [r0, #24]                                         \n"
+        " ldr r2, handler2_address_const                            \n"
+        " bx r2                                                     \n"
+        " handler2_address_const: .word prvGetRegistersFromStack    \n"
+    );
+}
+
+static void MemMang_Handler( void ) __attribute__( ( naked ) );
+void MemMang_Handler( void )
+{
+    __asm volatile
+    (
+        " tst lr, #4                                                         \n"
+        " ite eq                                                             \n"
+        " mrseq r0, msp                                                      \n"
+        " mrsne r0, psp                                                      \n"
+        " ldr r1, handler3_address_const                                      \n"
+        " bx r1                                                              \n"
+        " handler3_address_const: .word vHandleMemoryFault                    \n"
+    );
+}
+
+void BusFault_Handler( void )
+{
+    __asm volatile
+    (
+        " tst lr, #4                                                         \n"
+        " ite eq                                                             \n"
+        " mrseq r0, msp                                                      \n"
+        " mrsne r0, psp                                                      \n"
+        " ldr r1, handler4_address_const                                      \n"
+        " bx r1                                                              \n"
+        " handler4_address_const: .word vHandleMemoryFault                    \n"
+    );
+}
+
+void UsageFault_Handler( void )
+{
+    __asm volatile
+    (
+        " tst lr, #4                                                         \n"
+        " ite eq                                                             \n"
+        " mrseq r0, msp                                                      \n"
+        " mrsne r0, psp                                                      \n"
+        " ldr r1, handler5_address_const                                      \n"
+        " bx r1                                                              \n"
+        " handler5_address_const: .word vHandleMemoryFault                    \n"
+    );
+}
+
+void Debug_Handler( void )
+{
+    __asm volatile
```

Comment:
Not all of these are memory faults.

Reply:
nope

Reply:
yeah

Reply:
Rather than multiple of the same handler, can you just set each of the faults to use the same handler in the vector table?

Or build handlers that provide a bit more info upon failure for each class of fault.

Reply:
done